d-sandbox
<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 400px">
</div>

# Spark SQL
1. Run a SQL query
1. Create a DataFrame
1. Write same query using DataFrame transformations
1. Trigger computation with DataFrame actions
1. Convert between DataFrames and SQL

##### Methods
- SparkSession (<a href="https://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=sparksession#pyspark.sql.SparkSession" target="_blank">Python</a>/<a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/SparkSession.html" target="_blank">Scala</a>): `sql`, `table`
- DataFrame (<a href="https://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=dataframe#pyspark.sql.DataFrame" target="_blank">Python</a>/<a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">Scala</a>):
  - Transformations:  `select`, `where`, `orderBy`
  - Actions: `show`, `count`, `take`
  - Other methods: `printSchema`, `schema`, `createOrReplaceTempView`

In [0]:
%run ./Includes/Classroom-Setup-SQL

Datasets mounted and student environment set up

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Run a SQL query
Use `SparkSession` to run SQL

In [0]:
budgetDF = spark.sql("""
SELECT name, price
FROM products
WHERE price < 200
ORDER BY price
""")

View results in the returned DataFrame

In [0]:
budgetDF.show()

+--------------------+-----+
|                name|price|
+--------------------+-----+
|Standard Foam Pillow| 59.0|
|    King Foam Pillow| 79.0|
|Standard Down Pillow|119.0|
|    King Down Pillow|159.0|
+--------------------+-----+



In [0]:
display(budgetDF)

name price Standard Foam Pillow 59.0 King Foam Pillow 79.0 Standard Down Pillow 119.0 King Down Pillow 159.0

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Create a DataFrame
Use `SparkSession` to create a DataFrame from a table

In [0]:
productsDF = spark.table("products")
display(productsDF)

item_id name price M_PREM_Q Premium Queen Mattress 1795.0 M_STAN_F Standard Full Mattress 945.0 M_PREM_F Premium Full Mattress 1695.0 M_PREM_T Premium Twin Mattress 1095.0 M_PREM_K Premium King Mattress 1995.0 P_DOWN_S Standard Down Pillow 119.0 M_STAN_Q Standard Queen Mattress 1045.0 M_STAN_K Standard King Mattress 1195.0 M_STAN_T Standard Twin Mattress 595.0 P_FOAM_S Standard Foam Pillow 59.0 P_FOAM_K King Foam Pillow 79.0 P_DOWN_K King Down Pillow 159.0

Access schema of DataFrame

In [0]:
productsDF.printSchema()

root
 |-- item_id: string (nullable = true)
 |-- name: string (nullable = true)
 |-- price: double (nullable = true)



In [0]:
productsDF.schema

Out[99]: StructType(List(StructField(item_id,StringType,true),StructField(name,StringType,true),StructField(price,DoubleType,true)))

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Write same query with DataFrame transformations

In [0]:
budgetDF = (productsDF
  .select("name", "price")
  .where("price < 200")
  .orderBy("price")
)

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Trigger computation with DataFrame actions

In [0]:
budgetDF.count()

Out[101]: 4

In [0]:
budgetDF.take(2)

Out[102]: [Row(name='Standard Foam Pillow', price=59.0),
 Row(name='King Foam Pillow', price=79.0)]

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Convert between DataFrames and SQL

In [0]:
budgetDF.createOrReplaceTempView("budget")

In [0]:
display(spark.sql("SELECT * FROM budget"))

name price Standard Foam Pillow 59.0 King Foam Pillow 79.0 Standard Down Pillow 119.0 King Down Pillow 159.0

## ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Lab

1. Create a DataFrame from the `Event` table
1. Display DataFrame and inspect schema
1. Apply transformations to filter and sort `macOS` events
1. Count results and take first 5 rows
1. Create the same DataFrame using SQL query

##### Methods
- <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/SparkSession.html" target="_blank">SparkSession</a>: `sql`, `table`
- <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">DataFrame</a> transformations: `select`, `where`, `orderBy`
- <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">DataFrame</a> actions: `select`, `count`, `take`
- Other <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">DataFrame</a> methods: `printSchema`, `schema`, `createOrReplaceTempView`

### 1. Create a DataFrame from the `events` table
- Use SparkSession to create a DataFrame from the `events` table

In [0]:
# TODO
eventDF = spark.sql("""
SELECT event_name, device, items, event_timestamp
FROM events
""")

### 2. Display DataFrame and inspect schema
- Use methods above to inspect DataFrame contents and schema

In [0]:
# TODO
display(eventDF)

event_name device items event_timestamp warranty macOS List() 1593878946592107 press Windows List() 1593877011756535 add_item macOS List(List(null, M_STAN_T, Standard Twin Mattress, 595.0, 595.0, 1)) 1593878815459100 mattresses iOS List() 1593878809276923 mattresses Windows List() 1593878628143633 main Windows List() 1593878634344194 main iOS List() 1593877936171803 main macOS List() 1593876843215329 warranty Android List() 1593879213196400 main Windows List() 1593876713246514 original iOS List() 1593878170903989 main Linux List() 1593878036347579 down Android List() 1593879125815755 main Linux List() 1593878672173087 main macOS List() 1593876429415452 mattresses iOS List() 1593876687337581 premium macOS List() 1593877973962436 reviews Windows List() 1593876944661570 original iOS List() 1593877788141768 main iOS List() 1593877497207417 main Android List() 1593877271099453 reviews iOS List() 1593878356880855 main Android List() 1593879239050626 mattresses Windows List() 1593877608354608 cart Android List(List(null, M_STAN_F, Standard Full Mattress, 945.0, 945.0, 1)) 1593878899159806 cart Windows List(List(null, M_PREM_T, Premium Twin Mattress, 1095.0, 1095.0, 1)) 1593879003979444 main macOS List() 1593879039209043 main iOS List() 1593876687186216 faq macOS List() 1593878984168461 mattresses macOS List() 1593877076686136 main iOS List() 1593876831949693 pillows Windows List() 1593879086932085 main macOS List() 1593878642524459 main Linux List() 1593877977050669 pillows Windows List() 1593877512951677 mattresses Windows List() 1593878914256162 warranty Android List() 1593878620744974 pillows Windows List() 1593878751138630 foam iOS List() 1593878206468778 main Windows List() 1593878429326706 checkout Android List(List(null, M_STAN_K, Standard King Mattress, 1195.0, 1195.0, 1)) 1593879230950221 mattresses Android List() 1593876443386829 warranty Chrome OS List() 1593878823421232 main Windows List() 1593877041350735 mattresses Android List() 1593879240812006 main Chrome OS List() 1593878466349935 pillows macOS List() 1593879057039250 original macOS List() 1593878472750958 cart macOS List(List(null, M_STAN_K, Standard King Mattress, 1195.0, 1195.0, 1)) 1593877376501740 main Windows List() 1593878661450140 add_item Linux List(List(null, M_STAN_Q, Standard Queen Mattress, 1045.0, 1045.0, 1)) 1593878675942469 original macOS List() 1593877134114884 careers Windows List() 1593878120615516 warranty Android List() 1593878762928254 delivery Windows List() 1593878852225498 mattresses macOS List() 1593879062248629 main iOS List() 1593876967068480 main macOS List() 1593876568876937 press Linux List() 1593877572231604 original Android List() 1593878535553523 main Chrome OS List() 1593878981312640 mattresses Android List() 1593877756506929 add_item Android List(List(null, M_STAN_K, Standard King Mattress, 1195.0, 1195.0, 1)) 1593879138472812 main iOS List() 1593877868268844 main iOS List() 1593878487566059 warranty Windows List() 1593877853146507 main Windows List() 1593878946329881 cart Android List(List(null, M_PREM_T, Premium Twin Mattress, 1095.0, 1095.0, 1), List(null, P_FOAM_K, King Foam Pillow, 79.0, 79.0, 1)) 1593878826433752 main Chrome OS List() 1593876424543297 main Windows List() 1593876685393627 register iOS List(List(null, M_STAN_F, Standard Full Mattress, 945.0, 945.0, 1)) 1593879116208649 checkout macOS List(List(null, M_STAN_K, Standard King Mattress, 1195.0, 1195.0, 1)) 1593877871752279 main Windows List() 1593878639879302 reviews Android List() 1593879288072307 press iOS List() 1593878587026765 pillows macOS List() 1593876430929937 original Chrome OS List() 1593876953877757 add_item macOS List(List(null, M_STAN_Q, Standard Queen Mattress, 1045.0, 1045.0, 1)) 1593878872298444 premium Android List() 1593877636821563 mattresses Linux List() 1593878300547273 pillows macOS List() 1593877254607024 main macOS List() 1593877653737078 main Windows List() 1593876743752307 add_item Chrome OS List(List(null, M_STAN_K, Standard King Mattre

In [0]:
# TODO
eventDF.printSchema()

root
 |-- event_name: string (nullable = true)
 |-- device: string (nullable = true)
 |-- items: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- coupon: string (nullable = true)
 |    |    |-- item_id: string (nullable = true)
 |    |    |-- item_name: string (nullable = true)
 |    |    |-- item_revenue_in_usd: double (nullable = true)
 |    |    |-- price_in_usd: double (nullable = true)
 |    |    |-- quantity: long (nullable = true)
 |-- event_timestamp: long (nullable = true)



-sandbox
### 3. Apply transformations to filter and sort `macOS` events
- Filter for rows where `device` is `macOS`
- Sort rows by `event_timestamp`

<img alt="Hint" title="Hint" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-light-bulb.svg"/>&nbsp;**Hint:** Use single and double quotes in your filter SQL expression

In [0]:
# TODO
macDF =(eventDF
        .select("event_name", "device", "items", "event_timestamp")
        .where("'device' == 'macOS'")
        .orderBy("event_timestamp")
)

### 4. Count results and take first 5 rows
- Use DataFrame actions to count and take rows

In [0]:
# TODO
numRows = macDF.count()
rows = macDF.take(5)

In [0]:
numRows

Out[110]: 0

In [0]:
rows

Out[111]: []

-sandbox
##### <img alt="Best Practice" title="Best Practice" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-blue-ribbon.svg"/> Check your work

In [0]:
from pyspark.sql import Row

assert(numRows == 1938215)
assert(len(rows) == 5)
assert(type(rows[0]) == Row)

---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<command-2861581124099582> in <module>
      1 from pyspark.sql import Row
      2 
----> 3 assert(numRows == 1938215)
      4 assert(len(rows) == 5)
      5 assert(type(rows[0]) == Row)

AssertionError: 

### 5. Create the same DataFrame using SQL query
- Use SparkSession to run a sql query on the `events` table
- Use SQL commands above to write the same filter and sort query used earlier

In [0]:
# TODO
macSQLDF = spark.sql("""
SELECT event_name, device, items, event_timestamp
FROM events WHERE device=='macOS'
ORDER BY event_timestamp ASC
""")

display(macSQLDF)

event_name device items event_timestamp mattresses macOS List() 1592539216262230 mattresses macOS List() 1592539218667144 reviews macOS List() 1592539218939615 guest macOS List(List(null, M_STAN_F, Standard Full Mattress, 945.0, 945.0, 1)) 1592539225913937 main macOS List() 1592539226602157 original macOS List() 1592539229480664 main macOS List() 1592539231022266 guest macOS List(List(null, M_STAN_Q, Standard Queen Mattress, 1045.0, 1045.0, 1), List(null, P_FOAM_K, King Foam Pillow, 79.0, 79.0, 1)) 1592539231776200 delivery macOS List() 1592539232284730 main macOS List() 1592539233151229 shipping_info macOS List(List(null, M_STAN_F, Standard Full Mattress, 945.0, 945.0, 1)) 1592539235059340 mattresses macOS List() 1592539236295422 add_item macOS List(List(null, M_STAN_Q, Standard Queen Mattress, 1045.0, 1045.0, 1)) 1592539237432046 main macOS List() 1592539289834596 add_item macOS List(List(NEWBED10, P_FOAM_K, King Foam Pillow, 71.10000000000001, 79.0, 1)) 1592539295857466 main macOS List() 1592539299325832 main macOS List() 1592539302562339 mattresses macOS List() 1592539305168803 guest macOS List(List(NEWBED10, M_STAN_K, Standard King Mattress, 1075.5, 1195.0, 1)) 1592539309500570 add_item macOS List(List(NEWBED10, M_STAN_T, Standard Twin Mattress, 535.5, 595.0, 1)) 1592539310482088 register macOS List(List(NEWBED10, M_STAN_Q, Standard Queen Mattress, 940.5, 1045.0, 1)) 1592539331750583 shipping_info macOS List(List(NEWBED10, M_STAN_Q, Standard Queen Mattress, 940.5, 1045.0, 1)) 1592539339785249 mattresses macOS List() 1592539343571171 register macOS List(List(null, M_STAN_K, Standard King Mattress, 1195.0, 1195.0, 1)) 1592539346041388 pillows macOS List() 1592539347745795 original macOS List() 1592539367676338 press macOS List() 1592539383618319 cart macOS List(List(NEWBED10, M_PREM_Q, Premium Queen Mattress, 1615.5, 1795.0, 1)) 1592539387295729 main macOS List() 1592539388476666 original macOS List() 1592539388713269 add_item macOS List(List(null, M_STAN_F, Standard Full Mattress, 945.0, 945.0, 1)) 1592539389057282 mattresses macOS List() 1592539394685322 main macOS List() 1592539400000826 shipping_info macOS List(List(NEWBED10, M_STAN_Q, Standard Queen Mattress, 940.5, 1045.0, 1)) 1592539402672930 original macOS List() 1592539414743914 main macOS List() 1592539417279950 mattresses macOS List() 1592539417363593 shipping_info macOS List(List(NEWBED10, M_PREM_Q, Premium Queen Mattress, 1615.5, 1795.0, 1)) 1592539418536140 cart macOS List(List(NEWBED10, P_FOAM_K, King Foam Pillow, 71.10000000000001, 79.0, 1)) 1592539421435621 checkout macOS List(List(NEWBED10, P_FOAM_K, King Foam Pillow, 71.10000000000001, 79.0, 1)) 1592539425024905 register macOS List(List(NEWBED10, M_STAN_T, Standard Twin Mattress, 535.5, 595.0, 1)) 1592539430046139 main macOS List() 1592539430510397 delivery macOS List() 1592539433623588 add_item macOS List(List(NEWBED10, M_STAN_T, Standard Twin Mattress, 535.5, 595.0, 1)) 1592539439299177 checkout macOS List(List(null, M_STAN_K, Standard King Mattress, 1195.0, 1195.0, 1)) 1592539444746132 guest macOS List(List(null, M_STAN_K, Standard King Mattress, 1195.0, 1195.0, 1)) 1592539446975318 cc_info macOS List(List(NEWBED10, M_STAN_Q, Standard Queen Mattress, 940.5, 1045.0, 1)) 1592539453152225 cart macOS List(List(NEWBED10, M_PREM_Q, Premium Queen Mattress, 1615.5, 1795.0, 1)) 1592539453911179 mattresses macOS List() 1592539470197362 press macOS List() 1592539476111909 main macOS List() 1592539485558893 foam macOS List() 1592539493327941 original macOS List() 1592539494355700 mattresses macOS List() 1592539495439176 main macOS List() 1592539504294974 main macOS List() 1592539510772437 shipping_info macOS List(List(null, M_STAN_K, Standard King Mattress, 1195.0, 1195.0, 1)) 1592539513190152 main macOS List() 1592539515571793 main macOS List() 1592539516820101 mattresses macOS List() 1592539526219664 add_item macOS List(List(null, M_STAN_K, Standard King Mattress, 1195.0, 1195.0, 1)) 1592539528986263 premium mac

In [0]:
macSQLDF.take(5)

Out[120]: [Row(event_name='mattresses', device='macOS', items=[], event_timestamp=1592539216262230),
 Row(event_name='mattresses', device='macOS', items=[], event_timestamp=1592539218667144),
 Row(event_name='reviews', device='macOS', items=[], event_timestamp=1592539218939615),
 Row(event_name='guest', device='macOS', items=[Row(coupon=None, item_id='M_STAN_F', item_name='Standard Full Mattress', item_revenue_in_usd=945.0, price_in_usd=945.0, quantity=1)], event_timestamp=1592539225913937),
 Row(event_name='main', device='macOS', items=[], event_timestamp=1592539226602157)]

-sandbox
%md ##### <img alt="Best Practice" title="Best Practice" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-blue-ribbon.svg"/> Check your work
- You should only see `macOS` values in the `device` column  
- The fifth row should be an event with timestamp `1592539226602157`

### Classroom Cleanup

In [0]:
%run ./Includes/Classroom-Cleanup
